scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,268 @@
 #
 # Alternatively, the extauth-hook module can be called, which will
 # dispatch to the correct extauth-hook-<type>.py module automatically.
-
-from __future__ import print_function
+import abc
 import XenAPIPlugin
 import XenAPI
 import sys
 import os
-import getopt
 import tempfile
 import commands
-import syslog
+import logging
+import logging.handlers
+from enum import Enum
+
+
+def setup_logger():
+    logger = logging.getLogger()
+    logging.basicConfig(format='%(asctime)s %(levelname)s %(name)s %(funcName)s %(message)s', level=logging.DEBUG)
+    # Send to syslog local5, which will be redirected to xapi log /var/log/xensource.log
+    handler = logging.handlers.SysLogHandler(facility='local5', address='/dev/log')
+    std_err = logging.StreamHandler(sys.stderr)
+    logger.addHandler(handler)
+    logger.addHandler(std_err)
+
+
+setup_logger()
+logger = logging.getLogger(__name__)
+
+
+class ADBackend(Enum):
+    bd_pbis = 0
+    bd_winbind = 1
+
+
+class ADConfig(object):
+    def __init__(self, path, session, args, ad_enabled=True):
+        self._file_path = path
+        self._session = session
+        self._args = args
+        self._lines = []
+        self._backend = self._get_ad_backend(args)
+        self._ad_enabled = ad_enabled
+
+        if os.path.exists(self._file_path):
+            with open(self._file_path, 'r') as f:
+                self._lines = f.readlines()
+
+    def _get_ad_backend(self, args):
+        if  args.get("ad_backend", "winbind") == "pbis":
+            logger.debug("pbis is used as AD backend")
+            return ADBackend.bd_pbis
+
+        logger.debug("winbind is used as AD backend")
+        return ADBackend.bd_winbind
+
+    @abc.abstractmethod
+    def _apply_to_cache(self):
+        pass
 
-etc_pamd_sshd_start_boilerplate = """#%PAM-1.0
+    def apply(self):
+        self._apply_to_cache()
+        self._install()
+
+    def _install(self):
+        with tempfile.NamedTemporaryFile(prefix="extauth-", delete=False) as f:
+            lines = [line.encode() for line in self._lines]
+            f.writelines(lines)
```

Comment:
I'd use `join` here (and remove this line below `self._lines = ["{}\n".format(l) for l in lines[:-1]] + lines[-1:]`)

```suggestion
            f.write('\n'.join([l.encode() for l in self._lines])
```

Reply:
In you can remove occurrences of `\n` from a lot of places with this

Reply:
the point here is to  add `\n` to each line except the last line, can the suggestion match this?

Reply:
yes, see this example:

```
>>> '\n'.join(['a', 'b', 'c'])
'a\nb\nc'
```